### ISSUE TITLE
Review entity methods

### ISSUE THREAD

**Issue by Ichimonji10** [SatelliteQE/nailgun#41]
Several methods in module `nailgun.entities` do not map nicely to an underlying path. <ins>In addition, several methods do not have appropriate arguments.</ins> The methods exposed by the entity classes should be reviewed to make sure that they are appropriately named and placed on the most appropriate class.

Methods to review:

* [x] `ActivationKey.add_subscriptions`
* [x] `ActivationKey.content_override`
* [x] `AbstractDockerContainer.power`
* [x] `AbstractDockerContainer.logs`
* [x] `ContentViewVersion.promote`
* [x] `ContentView.publish`
* [x] `ContentView.set_repository_ids`
* [x] ~~`ContentView.available_puppet_modules`~~
* [x] `ContentView.add_puppet_module`
* [x] `ContentView.copy`
* [x] ~~`ForemanTask.poll`~~
* [x] `Organization.subscriptions`
* [x] `Organization.upload_manifest`
* [x] `Organization.delete_manifest`
* [x] `Organization.refresh_manifest`
* [x] `Organization.sync_plan`
* [x] ~~`Organization.list_rhproducts`~~ (EntitySearchMixin is used instead)
* [x] ~~`Permission.search`~~ (EntitySearchMixin is used instead)
* [x] `Product.list_repositorysets`
* [x] ~~`Product.fetch_rhproduct_id`~~ (EntitySearchMixin is used instead)
* [x] ~~`Product.fetch_reposet_id`~~ (EntitySearchMixin is used instead)
* [x] `Product.enable_rhrepo`
* [x] `Product.disable_rhrepo`
* [x] `Repository.sync`
* [x] ~~`Repository.fetch_repoid`~~ (EntitySearchMixin is used instead)
* [x] `Repository.upload_content`
* [x] `RHCIDeployment.add_hypervisors`
* [x] `RHCIDeployment.deploy`
* [x] `SyncPlan.add_products`
* [x] `SyncPlan.remove_products`

One example of an entity method that does not map nicely to an underlying path is [`Organization.upload_manifest`](http://nailgun.readthedocs.org/en/latest/api/nailgun.html#nailgun.entities.Organization.upload_manifest). This method POSTs to the following path:

    /katello/api/v2/organizations/<organization_id>/subscriptions/upload

However, the API documentation lists a second API path that can also be POSTed to:

    /katello/api/v2/subscriptions/upload

The latter approach is preferred. The method `Organization.upload_manifest` is not clearly related to any underlying path, but the method `Subscription.upload` is. There are also a bevy of problems with nested paths: they make it hard to determine what the canonical location of a resource it, it promotes a proliferation of the size of the API, and it makes it hard to manage special-case logic. (Special case logic is more easily managed when building a JSON payload than when building a URL.) Thus, `Organization.upload_manifest` should be dropped in favor of `Subscription.upload`. It could look like this:

```python
def upload(
        self, content, organization_id, repository_url=None,
        synchronous=True):
    """Upload a subscription manifest file.

    :param content: A file handle to a manifest file. In other words,
        the object returned by `open('path/to/file', 'rb')`.
    :param int organization_id: An organization ID.
    :param str repository_url: An optional repository URL.
    :param bool synchronous: What should happen if the server returns an
        HTTP 202 (accepted) status code? Wait for the task to complete if
        ``True``. Immediately return JSON response otherwise.
    :returns: The server's response, with all JSON decoded.
    :rtype: dict
    :raises: ``nailgun.entity_mixins.TaskTimedOutError`` if an HTTP 202
        response is received, ``synchronous is True`` and the task
        completes with any result other than "success".

    """
    data = {'organization_id': organization_id}
    if repository_url is not None:
        data['repository_url'] = repository_url
    response = client.post(
        self.path('upload'),
        data,
        auth=self._server_config.auth,
        files={'content': content},
        verify=self._server_config.verify,
    )
    response.raise_for_status()

    # Poll a task if necessary, then return the JSON response.
    if synchronous is True and response.status_code == httplib.ACCEPTED:
        return ForemanTask(
            self._server_config,
            id=response.json()['id']
        ).poll()
    return response.json()
```

**Comment by Ichimonji10** [comment]
In addition to verifying that each entity method maps to an appropriate underlying API path, the arguments to each entity method should be verified. How so? An an example, take a look at these two method signatures:

    def add_subscriptions(self, params):
    def content_override(self, content_label, value):

The first method accepts a single parameter, `params`. `params` is a dict that is inserted directly into an HTTP PUT call:

```python
response = client.put(
    self.path('add_subscriptions'),
    params,
    auth=self._server_config.auth,
    verify=self._server_config.verify,
)
```

The second method accepts two parameters. These parameters are placed into an HTTP PUT call one-by-one:

```python
response = client.put(
    self.path('content_override'),
    {'content_override': {
        'content_label': content_label,
        'value': value,
    }},
    auth=self._server_config.auth,
    verify=self._server_config.verify,
)
```

Which strategy is better? I think that the latter strategy is sinificantly better. One of the aims of NailGun is to shield developers from the idiosyncracies of the API, and one of those idiosyncracies is the uneven usage of nested dicts. The latter method allows a developer to pass in information and now worry about the layout of the JSON payload. In addition, the latter method allows each individual parameter to be documented individually. This means that a dev can read relevant method documentation at http://nailgun.readthedocs.org/en/latest/api/nailgun.html#module-nailgun.entities, instead of being forced to trawl through the API documentation that ships with Satellite.

What's the take-away? In addition to reviewing method placement and naming, method arguments should also be reviewed.

**Comment by omaciel** [comment]
Nice write up! I totally grok the suggestion that explicitly stating the arguments makes it easier to read and understand the code.

**Comment by Ichimonji10** [comment]
Method `Organization.subscription` should probably be dropped in favor of method `Subscription.search`. Here's what the method signature could look like:

    def search(self, organization, system, activation_key):

You could use this method to find all subscriptions belonging to a particular organization like so:

```python
>>> subscriptions = entities.Subscription.search(
...    organization=entities.Organization(id=5)
... )
>>> type(subscriptions)
list
>>> isinstance(subscriptions[0], entities.Subscription)
True
```

Creating and passing in an object every time you want to call `Subscription.search` is cumbersome. In response, I think that it's reasonable to offer the same shortcut as is offered by `EntityCreateMixin.create`:

```python
>>> subscriptions = entities.Subscription.search(organization=5)
>>> type(subscriptions)
list
>>> isinstance(subscriptions[0], entities.Subscription)
True
```

As you might have figured out, this particular fix depends on #11.

**Comment by elyezer** [comment]
What about accepting both:

```python
>>> subscriptions = entities.Subscription.search(organization_id=5)
>>> type(subscriptions)
list
>>> isinstance(subscriptions[0], entities.Subscription)
True
```

and

```python
>>> subscriptions = entities.Subscription.search(organization=org_instance)
>>> type(subscriptions)
list
>>> isinstance(subscriptions[0], entities.Subscription)
True
```

**Comment by Ichimonji10** [comment]
@elyezer, I agree.

I'm afraid I wasn't clear in my previous post. I was suggesting exactly what you've suggested: that the method accept either an entity object or an ID. You made the suggestion much more clear, though. :smile:

**Comment by elyezer** [comment]
@Ichimonji10 thanks for the clarification. accepting both seems to be the right and more useful thing to do :smile:

**Comment by Ichimonji10** [comment]
[This comment](https://github.com/SatelliteQE/nailgun/issues/41#issuecomment-94506248) discusses an approach for making method signatures. It provides these two method signatures as an example:

```python
def add_subscriptions(self, params):
def content_override(self, content_label, value):
```

That comment then goes on to discuss why the latter method signature is preferred. In short:

* The latter method is more discoverable. A user can simply look at the possible arguments, or even the docstring for that method, and have an idea about what information the method requires.
* The latter method allows NailGun to shield users from design quirks (such as the API requiring a nested hash) or bugs (such as the API redundantly requiring an entity ID).

There's one challenge in this approach that has not yet been discussed. The challenge is wrapping calls such as `PUT /fusor/api/v21/deployments/:id/deploy`. [Currently,](https://github.com/SatelliteQE/nailgun/pull/98#issuecomment-107681829)

> the deploy method only accepts one optional parameter, but could be expanded to more in future builds. Also, the update method can be used to update any deployment value

In other words, explicitly listing parameters when those parameters are *optional* and *variable* is problematic. It doesn't make sense to write this:

```python
def deploy(foo=None, bar=None, biz=None):
```

…given that those arguments may change in the future, and given that passing `foo=null` to the server is different than not passing `foo` at all.

The currently adopted solution is to just accept a dict of parameters that are passed straight to the server. However, accepting a dict of parameters and handing them straight to the server is not ideal for all the reasons summarized at the top of this post. It's also not ideal because we're adopting two very different approaches for passing in arguments to methods: passing in a dict and explicitly naming arguments.

One solution is to accept a list of kwargs for methods such as `deploy()`:

```python
def deploy(**kwargs):
```

This has the advantage of letting the call syntax be similar for all helper methods.

**Comment by elyezer** [comment]
I like the `**kwargs` approach. Then in the docstring we can tell the user what are the required expected arguments and what are optional.

**Comment by oshtaier** [comment]
To summarize all statements above and define goal for current issue:

1) Move necessary method (e.g. 'upload_manifest') to corresponding entities (like from 'Organization' path to 'Subscription') if applicable for that path.
2) Use `**kwargs` approach for all helper methods call syntax.
~~3) Statements about search approach mentioned above is already covered in EntitySearchMixin~~

Does it sound right?

**Comment by Ichimonji10** [comment]
1. Right.
2. Maybe. Using kwargs is problematic where the keywords expected by Satellite conflict with keywords used for other purposes (e.g. "self", "synchronous").

**Comment by Ichimonji10** [comment]
Discussion
==========

There's at least three approaches we could take in designing the signatures of
NailGun's helper methods:

1. Accept arguments individually and explicitly list each possible argument.
2. Accept arguments individually via `**kwargs`.
3. Accept arguments together as a dict.

As an example of each, consider `Subscription.upload`:

    def upload(self, content, organization, repository=None, synchronous=True):
    def upload(self, synchronous=True, **kwargs):
    def upload(self, payload, synchronous=True):

Which should be used? To answer that question, I use the following rubric:

* Can a user look at a method signature and its docstring and quickly start
  using the method?
* Can a user look at a method usage and intuit what is happening?
* Can a developer easily implement a method?
* Can a developer easily maintain a method over time?

Approach 1
----------

Recap:

> Accept arguments individually and explicitly list each possible argument.
>
>     def upload(self, content, organization, repository=None, synchronous=True):

### Signature & Docstring

Each possible argument is listed out. Furthermore, arguments required by the
server are also required by the method, and arguments not required by the server
are not required by the method. This helps.

There is no obvious distinction between the `synchronous` argument and other
arguments, which hurts.

### Usage

Consider these example method calls:

    upload(my_file_handle, my_org)
    upload(my_file_handle, my_org, my_repository)
    upload(my_file_handle, my_org, my_repository, False)

Each individual argument may be passed by position for conciseness or by keyword
for clarity. This helps.

The distinction between the `synchronous` argument and other arguments is not
obvious, which hurts.

The form of the payload sent to the server is not obvious. This hurts.

### Implementation

The developer must consult with the API documentation and discover where to
place each individual method in the request payload. This hurts.

The keywords required by the server may conflict with reserved words such as
"self", "synchronous", "type", "if", etc. This hurts.

### Maintenance

This method is *extremely* hard to maintain.

* Imagine that the structure of arguments accepted by the server changes. In
  this case, a version comparison must be done in the method and each possible
  layout must be listed.
* Imagine that the set of arguments accepted by the server changes. In this
  case, some painful hack must be performed, such as defining several nearly
  identical methods and binding one of them to an object at instantiation-time
  based on the server software version.

Approach 2
----------

Recap:

> 2. Accept arguments individually via `**kwargs`.
>
>     def upload(self, synchronous=True, **kwargs):

### Signature & Docstring

Possible arguments are not listed out. Instead, users must consult with the
docstring or API guide to see what arguments are acceptable. This hurts.

There is an obvious distinction between the `synchronous` argument and other
arguments, which helps.

`**kwargs` magic is used, which hurts.

### Usage

Consider these example method calls:

    upload(content=my_file_handle, organization_id=org.id)
    upload(
        content=my_file_handle,
        organization_id=org.id,
        repository_url=repo_url,
        synchronous=False,
    )

Each individual argument must be passed by keyword, which hurts.

The distinction between the `synchronous` argument is not obvious. Even worse,
all arguments are always passed by keyword, which means that arguments are
likely to be re-ordered, further blurring the distinction between `synchronous`
and other arguments. This hurts.

The form of the payload sent to the server is not obvious. This hurts.

### Implementation

The developer may blindly pass `**kwargs` to whatever `client.*` method is being
used, which helps.

The keywords required by the server may conflict with reserved words such as
"self", "synchronous", "type", "if", etc. This hurts.

### Maintenance

No changes are required if the set of arguments changes. No changes are required
if the semantics of arguments changes, such as which are required or optional.
This helps.

Clients may immediately use new features without any changes on our part. This
helps.

Approach 3
----------

Recap:

> 3. Accept arguments together as a dict.
>
>     def upload(self, payload, synchronous=True):

### Signature & Docstring

Possible arguments are not listed out. Instead, users must consult with the
docstring or API guide to see what arguments are acceptable. This hurts.

There is an obvious distinction between the `synchronous` argument and other
arguments, which helps.

### Usage

Consider these example method calls:

    upload({
        'content': my_handle,
        'organization_id': org.id,
        'repository_url': repo_url,
    })
    upload(payload, False)
    upload(payload, synchronous=False)

Each argument may be passed when the method is called or arguments may be
assembled in to a dict before-hand. Furthermore, all arguments may be passed
either by position or keyword. As a result, developers may choose whatever form
makes the most sense for their context. This helps.

The distinction between the `synchronous` argument and other arguments is
obvious. This helps.

The form of the payload sent to the server is obvious. This helps.

The methods provided by the mixins (`create*`, `read*`, `update*`, `delete*`,
`search*`) all list keywords as arguments, which is different from what is being
done here. That said, those methods are significantly different from helper
methods. Mixin methods use object state extensively, read payload arguments from
the object they are called on, and make use of call trees for easy overloading
and tweaking. In contrast, helper methods are one-off and make little use of
object state (mainly `path()`). As a result, letting the helper methods take a
different form from the mixin methods may actually be good.

### Implementation

The developer may blindly pass `payload` to whatever `client.*` method is being
used, which helps.

The keywords required by the server will never conflict with reserved words such
as "self", "synchronous", "type", "if", etc. This helps.

### Maintenance

No changes are required if the set of arguments changes. No changes are required
if the semantics of arguments changes, such as which are required or optional.
This helps.

Clients may immediately use new features without any changes on our part. This
helps.

Analysis
--------

Approach 1 is inadvisable. It will be a hassle to implement and a nightmare to
maintain.

Both approach 1 and 2 suffer from several issues resulting from their use of
keyword arguments:

* The `synchronous` argument is not clearly separate from arguments passed to
  the server. If any other arguments are added for some reason, they also are
  not clearly separated from arguments passed to the server.
* The arguments required by the server may conflict with reserved words: "self",
  "type", "synchronous", "def", etc. We work around this in the
  `Entity.__init__` method by clever use of sub-classing, which is doable. But
  working around those same issues in numerous once-off methods too will make
  implementation and maintenance harder.
* The form of the payload sent to the server is unclear. This is fine for the
  mixins — they provide `*_payload` methods, and an interested user may call one
  of those methods to see what is being sent to the server. But the convenience
  methods do not allow for that level of inspection, meaning that users will
  need to read the source code (or something even more laborious) to understand
  exactly what's going on.

One point discussed above is that using keyword arguments lets the helper
methods be similar to other methods in NailGun, such as `Entity.__init__` and
`EntityCreateMixin.create`. However, as discussed above:

> helper methods are one-off and make little use of object state (mainly
> `path()`). As a result, letting the helper methods take a different form from
> the mixin methods may actually be good.

Conclusion
----------

Use approach 3. All methods should have a signature in one of the following
forms:

    def helper_method(self, payload):
    def helper_method(self, payload, synchronous=True):
    def helper_method(self, payload, synchronous=True, …):

Be warned that these changes will produce *massive* breakage in Robottelo.

Miscellaneous Notes
-------------------

Use the shortest available path for performing any given task, where length is
measured in path segments. For example, given these two API calls:

    POST /katello/api/v2/organizations/:organization_id/subscriptions/upload
    POST /katello/api/v2/subscriptions/upload

Make the helper method call the latter path. Require that the user pass in an
organization ID in the arguments dict. I know, I know, it's inconvenient in many
cases. But being convenient means preventing the user from making certain types
of method calls. For example, imagine that the user wants to do this:

    my_subscription.upload({'organization_id': org.id, …})

What happens if `my_subscription.organization` is already set? Will that win
out, or will the argument passed in win out? What if no organization is set on
the object and none is passed in. Will an exception be raised when some bit of
"convenience" code tries to set the `organization_id` argument for us? What if a
quality engineer specifically wants to pass no organization ID. How can the
quality engineer do that?

It's much better to act dumb and let the user be explicit about what they want.
"When in doubt, leave it out!"

**Comment by omaciel** [comment]
Thank you for the write up @Ichimonji10 ! I too tend to lean toward approach 3 for all the reasons you already hashed out.

**Comment by abalakh** [comment]
@Ichimonji10, thanks for explanations. One question from me:
What if method accepts only 1 argument and is unlikely to accept anything else in future? E.g. `SyncPlan.add_products()` - it only accepts a list of product ids to add and that's it. Or `Repository.upload_content()` - accepts handle of file to upload. Should we skip them from changes (as there are probably not so many benefits in additional dict for them) or it's better to change them anyway - to keep everything in the same style?

**Comment by Ichimonji10** [comment]
@abalakh Please give everything the same style. Doing so gives nailgun a lower conceptual weight. I'd like it if a user could learn how to use one of the helper method and then trust that every single helper method is written the same.

**Comment by abalakh** [comment]
One more thing - sometimes one of arguments is required in path. For example,
```python
    def disable_rhrepo(self, base_arch,
                       release_ver, reposet_id, synchronous=True):
```
`reposet_id` from here is used in request path later. I can either 
a) leave reposet_id required:
```python
    def disable_rhrepo(self, reposet_id, payload, synchronous=True):
```
but it's move away from the style.

or
b) move everything to payload, and raise an exception if `reposet_id` wasn't found in it:
```python
    def disable_rhrepo(self, payload, synchronous=True):
        if 'reposet_id' not in payload:
            raise(...)
```
but then we're loosing following benefits:
>No changes are required if the set of arguments changes. No changes are required if the semantics of arguments changes, such as which are required or optional. This helps.
>Clients may immediately use new features without any changes on our part. This helps.

In addition `reposet_id` shouldn't actually be in payload (not a big deal, can `pop()` it out, but that looks like workaround for me).
Any suggestions/better ideas?

**Comment by Ichimonji10** [comment]
@abalakh Drop the `enable_rhrepo` method. Define a new entity:

```python
class RepositorySet(Entity, …):

    def __init__(self, server_config=None, **kwargs):
        _check_for_value('product', kwargs)
        self._fields = {…}
        super(RepositorySet, self).__init__(server_config, **kwargs)
        self._meta = {
            'api_path': '{0}/repository_sets'.format(self.product.path()),
        }

    def enable(self, payload, synchronous=True):
        response = client.post(…)
        return _handle_response(response, self._server_config, synchronous)
```

**Comment by oshtaier** [comment]
Regarding that set of methods:
```
 Organization.subscriptions
 Organization.upload_manifest
 Organization.delete_manifest
 Organization.refresh_manifest
```
It was mentioned in the previous comments that it will be good to drop `Organization.upload_manifest` method in favor of `Subscription.upload`. And, that idea seems really great and must be implemented in that way if we have proper API path like this:
`POST /katello/api/subscriptions/upload` for all methods, but that is not true at that moment:

Organization.subscriptions:
`GET /katello/api/subscriptions/:id`
Organization.upload_manifest:
`POST /katello/api/subscriptions/upload`
Organization.delete_manifest:
`POST /katello/api/organizations/:organization_id/subscriptions/delete_manifest` (*Only this*)
Organization.refresh_manifest:
`PUT /katello/api/organizations/:organization_id/subscriptions/refresh_manifest` (*Only this*)

As we do not have proper implementation for all paths for mentioned functionality from `Subscription` entity point of view, we should proceed with old approach (left all these methods in `Organization` entity). Otherwise, we will have things like that:
```
subscription.upload_manifest()
org.delete_manifest()
```
which is obviously not good idea, especially from tool user perspective

**Comment by oshtaier** [comment]
Removing `Organization.list_rhproducts` as same functionality can be achieved using our new powerful EntitySearchMixin. For example:
```
org = entities.Organization().create()
product = entities.Product(organization=org).create()
product2 = entities.Product(organization=org).create()
results = entities.Product(organization=org).search({'organization'}, query={'per_page': 50, 'order': 'name DESC'})
```
will return list of products for particular organization

**Comment by elyezer** [comment]
I agree with the approach of passing payload as a dict, that will make all helper methods have the same signature. The downside is that the user will need to know the required fields to be passed, but we can minimize this with some methods that can help the user to check that information. Will try to expand the suggestion on a separated issue.

**Comment by sthirugn** [comment]
Yes, I agree with approach 3 as well with the understanding that this may need a few util methods in robottelo.

**Comment by abalakh** [comment]
Ok, so as per now:
* all the helper funcs are updated with approach 3.
* API paths were not touched, unfortunately. 2 paths could be changed - for Subscriptions and SyncPlans, but short paths for both of them are missing some required functions (e.g. create for SyncPlan and delete for subscription manifest), making it impossible to completely switch to the short ones.
* Helper funcs, purpose of which was to search some entity are removed and EntitySearchMixin is used instead.

**Comment by Ichimonji10** [comment]
@abalakh That makes perfect sense. Thanks for the feedback.

@sthirugn @elyezer Yeah, it's a real bugger that the user will have to know exactly what to send to the server. :disappointed:

@oshtaier Believe it or not, it's doable to drop the `Organization.delete_manifest` method in favor of a `Subscription.delete_manifest` method. The same is true for the other manifest-related methods, too. The diff would look like this (snippet):

```diff
-class Subscription(Entity):
+class Subscription(
+        Entity,
+        EntityReadMixin,
+        EntitySearchMixin):
     """A representation of a Subscription entity."""
 
     def __init__(self, server_config=None, **kwargs):
-        # NOTE: When making an HTTP POST call, `pool_uuid` must be renamed to
-        # `id`. This logic can be packed in to create_payload().
         self._fields = {
             'activation_key': entity_fields.OneToOneField(ActivationKey),
-            'pool_uuid': entity_fields.StringField(),
+            'organization': entity_fields.OneToOneField(Organization),
             'quantity': entity_fields.IntegerField(),
             'subscriptions': entity_fields.OneToManyField(Subscription),
             'system': entity_fields.OneToOneField(System),
         }
         self._meta = {
-            'api_path': 'katello/api/v2/subscriptions/:id',
+            'api_path': 'katello/api/v2/subscriptions',
             'server_modes': ('sat', 'sam'),
         }
         super(Subscription, self).__init__(server_config, **kwargs)
 
+    def path(self, which=None):
+        if which in (
+                'delete_manifest',
+                'manifest_history',
+                'refresh_manifest',
+                'upload'):
+            _check_for_value('organization', self.get_values())
+            return '{0}/subscriptions/{1}'.format(
+                self.organization.path('self'),  # pylint:disable=no-member
+                which
+            )
+        return super(Subscription, self).path(which)
+
+    def delete_manifest(self, synchronous=True):
+        response = client.post(
+            self.path('delete_manifest'),
+            **self._server_config.get_client_kwargs()
+        )
+        return _handle_response(response, self._server_config, synchronous)
+
+    def manifest_history(self, synchronous=True):
+        response = client.get(
+            self.path('manifest_history'),
+            **self._server_config.get_client_kwargs()
+        )
+        return _handle_response(response, self._server_config, synchronous)
+
+    def refresh_manifest(self, synchronous=True):
+        response = client.put(
+            self.path('refresh_manifest'),
+            **self._server_config.get_client_kwargs()
+        )
+        return _handle_response(response, self._server_config, synchronous)
+
+    def upload(self, synchronous=True):
+        response = client.put(
+            self.path('upload'),
+            **self._server_config.get_client_kwargs()
+        )
+        return _handle_response(response, self._server_config, synchronous)
+
```

**Comment by Ichimonji10** [comment]
Thanks for the hard work, @oshtaier and @abalakh. I'd like to do a bit more work in this branch before merging and declaring #41 fixed. Here's the list (sorry about formatting).

Modify:

* `ActivationKey.content_override` Pass payload directly instead of wrapping in a dict.
* (debatable) `Repository.upload_content`: Normalize file upload logic. Allow a
  separate "files" parameter to be passed in. This lets us use the full power of
  requests, including piping data directly to the method instead of reading it
  from disk.

Drop:

* `ContentView.add_puppet_module` Drop in favor of `ContentViewPuppetModule.create`.
* `Organization.sync_plan` Drop in favor of `SyncPlan(organization=org).create()`
* `RHCIDeployment.add_hypervisors` Drop in favor of `RHCIDeployment.update`.
* `RepositorySet.list_all` Drop in favor of `RepositorySet.search`.

Move:

* `Organization.delete_manifest` → `Subscription.delete_manifest`
* `Organization.refresh_manifest` → `Subscription.refresh_manifest`
* `Organization.subscriptions` → `Subscription(organization=org).search()`
* `Organization.upload_manifest` → `Subscription.upload`

**Comment by oshtaier** [comment]
@Ichimonji10 To be honest, it is hard for my understanding to name:
```
POST /katello/api/organizations/:organization_id/subscriptions/upload
PUT /katello/api/organizations/:organization_id/subscriptions/refresh_manifest
POST /katello/api/organizations/:organization_id/subscriptions/delete_manifest
```
as a subscription methods. From point of logic such statement seems pretty incorrect as for me.
If we use:
```
POST /katello/api/subscriptions/upload
...
```
that will be fine, but not like above.
As a temporary solution it might work, but not as permanent as it can confuse a lot.

**Comment by Ichimonji10** [comment]
@oshtaier There are a lot of entities that are available at more than one path. These API calls are synonymous:

* GET /docker/api/v2/containers
* GET /docker/api/v2/compute_resources/:compute_resource_id/containers

Similarly, these four paths are synonymous:

* GET /api/domains
* GET /api/subnets/:subnet_id/domains
* GET /api/locations/:location_id/domains
* GET /api/organizations/:organization_id/domains

These two paths are also synonymous:

* POST /katello/api/host_collections
* POST /katello/api/organizations/:organization_id/host_collections

Whenever you see "foo/:foo_id" in a path, you can mentally drop that path segment. That piece of information may somehow be necessary when making an API call (either as part of the URL path, the URL query or as a JSON payload) but it doesn't identify which entity is being worked with.

**Comment by oshtaier** [comment]
@Ichimonji10 Okay. That things is pretty obvious for me. Anyway, I thought if we have two 'good' (short) and two 'bad' (long) paths - it will be better to wait when we have 4 'good' ones.

I am just wondering whether we have similar situation before. Something like:
Entity:
`domains`
Path:
```
GET /api/subnets/:subnet_id/domains
```
or
Entity:
`containers`
Path:
```
GET /docker/api/v2/compute_resources/:compute_resource_id/containers
```
From my observation we always use paths like `/api/**entity_name**`

**Comment by oshtaier** [comment]
But your approach is a best one from user perspective as I don't think that someone care about all these paths issues.

**Comment by Ichimonji10** [comment]
@oshtaier That's basically it. We have to do some gymnastics on the back end to make everything work nicely. But the end result is that a user gets to work with a "Subscription" entity without worrying about which of the several API paths are used for a given call. On the back end, everything should work pretty nicely going forward, too. Let's say that Satellite deprecates all the long paths in favor of short paths, or deprecates all the short paths in favor of long paths. That's no problem - we can use a little versioning logic to deal with the change and the user is never the wiser.

**Comment by oshtaier** [comment]
@Ichimonji10 Thanks
Back to robottelo.
Previously we had:
`entities.Organization(name=org_name).create().upload_manifest(path=manifests.clone())`
after moving to 'Subscription' approach we will need to create one more entity, right?
```
org = entities.Organization(name=org_name).create()
subscription = entities.Subscription(organization=org).create()
subscription.upload(path=manifests.clone())
```
Don't we introduce redundant code/operations with our improvements?

**Comment by Ichimonji10** [comment]
@oshtaier Here's sample code:

```python
org = entities.Organization(name=org_name).create()
sub = entities.Subscription()
sub.delete_manifest({'organization_id': org.id})
sub.manifest_history({'organization_id': org.id})
sub.refresh_manifest({'organization_id': org.id})
with open('manifest.zip') as manifest:
    sub.upload({'organization_id': org.id}, manifest)
```

That last code sample could also be written like so:

```python
with open('manifest.zip') as manifest:
    sub.upload({'organization_id': org.id, 'content': manifest})
```

But the latter approach has some nasty side effects. Consider:

```python
with open('manifest.zip') as manifest:
    payload = {'organization_id': org.id, 'content': manifest})
    sub.upload(payload)
    sub.upload(payload)  # Fails! 'content' not in payload
```
